### PR TITLE
Add KeyManagedEncryptedEnv API

### PR DESCRIPTION
**Pull request by yiwu-arbug**
Add `KeyManagedEncryptedEnv` API and related `EncryptionKeyManager` trait.

### CODE REVIEW

File: librocksdb_sys/Cargo.toml [hunk 1]
```diff
@@ -8,6 +8,7 @@ links = "rocksdb"
 [dependencies]
 libc = "0.2.11"
 libtitan_sys = { path = "libtitan_sys" }
+openssl-sys = { version = "0.9.54", features = ["vendored"] }
```
Comment (overvenus) [inline librocksdb_sys/Cargo.toml]: Instead of `vendored` by default, could you define a feature?

File: src/encryption.rs [hunk 2]
```diff
@@ -0,0 +1,810 @@
+// Copyright 2014 Tyler Neely
```
Comment (overvenus) [inline src/encryption.rs]: Is it correct?

File: src/rocksdb.rs [hunk 3]
```diff
@@ -2525,6 +2526,23 @@ impl Env {
         Env::new_ctr_encrypted_env(Arc::new(Env::default()), ciphertext)
     }
 
+    // Create an encrypted env that accepts an external key manager.
+    pub fn new_key_managed_encrypted_env(
+        base_env: Arc<Env>,
+        key_manager: DBEncryptionKeyManager,
```
Comment (overvenus) [inline src/rocksdb.rs]:
```suggestion
        key_manager: Arc<dyn EncryptionKeyManager>,
```

What about `Arc<dyn EncryptionKeyManager>` and making `DBEncryptionKeyManager` unexported?

File: librocksdb_sys/Cargo.toml (hunk 1 above)
Comment (yiwu-arbug) [reply librocksdb_sys/Cargo.toml]: tikv is setting "vendored" by default for openssl: https://github.com/tikv/tikv/blob/master/Cargo.toml#L82

what's the use case when we don't want static link?

File: src/encryption.rs (hunk 2 above)
Comment (yiwu-arbug) [reply src/encryption.rs]: Fixed.

File: src/rocksdb.rs (hunk 3 above)
Comment (yiwu-arbug) [reply src/rocksdb.rs]: good suggestion. fixed.

File: librocksdb_sys/Cargo.toml (hunk 1 above)
Comment (overvenus) [reply librocksdb_sys/Cargo.toml]: Other projects may use this repo, and they may want to build without openssl.

**Comment by yiwu-arbug** [comment]
> I suggest to make openssl optional and add a feature to enable it. Rest LGTM

updated. I thought you mean add feature to enable static link.

File: src/encryption.rs [hunk 4]
```diff
@@ -0,0 +1,798 @@
...
+                ))
+                .unwrap()
+                .into_raw(),
+            )
+        },
+    }
+}
+
+impl DBEncryptionKeyManager {
+    pub fn new(key_manager: Arc<dyn EncryptionKeyManager>) -> DBEncryptionKeyManager {
```
Comment (overvenus) [inline src/encryption.rs]: Could you move it next to `impl Drop for DBEncryptionKeyManager`?

File: src/encryption.rs [hunk 5]
```diff
@@ -0,0 +1,798 @@
...
+        // The size of the raw pointer is of 16 bytes, and it doesn't fit in a C-style pointer.
+        // Allocate a buffer of 16 bytes to store as ctx in C code.
+        let ptr = Arc::into_raw(key_manager);
+        let instance = unsafe {
+            let ctx = libc::malloc(mem::size_of_val(&ptr));
```
Comment (overvenus) [inline src/encryption.rs]: How about `Box::new(Arc<dyn EncryptionKeyManager>)`?

File: src/encryption.rs [hunk 6]
```diff
@@ -0,0 +1,798 @@
...
+            *(ctx as *mut *const dyn EncryptionKeyManager) = ptr;
+            crocksdb_ffi::crocksdb_encryption_key_manager_create(
+                ctx,
+                encryption_key_manager_destructor,
```
Comment (overvenus) [inline src/encryption.rs]: What about keep a pointer of `ctx` in `DBEncryptionKeyManager` and clean the `ctx during `Drop` as the `ctx` is created in the Rust side?

File: src/encryption.rs [hunk 7]
```diff
@@ -0,0 +1,798 @@
...
+        Err(err) => {
+            return unsafe {
+                libc::strdup(
+                    CString::new(format!(
+                        "Encryption key manager encounter non-utf8 file name: {}",
+                        err
+                    ))
+                    .unwrap()
+                    .into_raw(),
+                )
```
Comment (overvenus) [inline src/encryption.rs]: It looks like leaking. 
1. CString::into_raw() consumes the CString and transfers ownership of the string to a C caller.
2. strdup allocates and duplicates the string of into_raw().

So 1 is leaked.

Comment (yiwu-arbug) [reply src/encryption.rs]: The pointer is received by C caller and deleted. See, for example, crocksdb_encryption_key_manager_impl_t::RenameFile in c.cc.

File: src/encryption.rs (hunk 6 above)
Comment (yiwu-arbug) [reply src/encryption.rs]: `DBEncryptionKeyManager` is just a wrapper holding a c++ shared_ptr. It is dropped immediately after caller calling `Env::new_key_managed_encrypted_env()`. The cleanup of `ctx` happen in destructor of c++ `KeyManagedEncryptedEnv` when it destruct the shared_ptr.

File: src/encryption.rs (hunk 7 above)
Comment (yiwu-arbug) [reply src/encryption.rs]: good catch! fixed.

File: src/encryption.rs (hunk 6 above)
Comment (yiwu-arbug) [reply src/encryption.rs]: Cleanup of  the `Arc` finally happens in `encryption_key_manager_destructor`.

File: src/encryption.rs [hunk 8]
```diff
@@ -0,0 +1,729 @@
...
+    fn new_file(&self, fname: &str) -> Result<FileEncryptionInfo>;
+    fn delete_file(&self, fname: &str) -> Result<()>;
+    fn link_file(&self, src_fname: &str, dst_fname: &str) -> Result<()>;
+    fn rename_file(&self, src_fname: &str, dst_fname: &str) -> Result<()>;
+}
+
+// Copy rust-owned error message to C-owned string. Caller is responsible to delete the result.
+fn copy_error<T: Into<Vec<u8>>>(err: T) -> *const c_char {
+    let cstr = CString::new(err).unwrap();
+    unsafe { libc::strdup(cstr.as_ptr()) }
```
Comment (overvenus) [inline src/encryption.rs]:
```suggestion
    CString::new(err).into_raw()
```

To save an extra allocation, we can just use `CString::into_raw()`.

Comment (yiwu-arbug) [reply src/encryption.rs]: We cannot. The document of `CString::into_raw()` says the result is not safe to be freed in C.

**Pull request merged**